### ISSUE TITLE
PC1 with FIL_PROOFS_USE_MULTICORE_SDR=1 may be OOM killed despite swap available

### ISSUE THREAD

**Issue by neondragon** [filecoin-project/rust-fil-proofs#1410]
When one lotus-worker process runs two PC1 tasks in parallel, and FIL_PROOFS_USE_MULTICORE_SDR=1, I (Slack NeonixAF f019551) have noticed an issue where the process is OOM killed around 20 minutes in to running the tasks, when lotus-worker's physical memory usage reaches 128GB. 256GB swap was available but mostly unused. I've thought this was due to lotus-worker locking pages in memory which the Linux kernel refused to evict, however, I've been told by other miners that they are able to use FIL_PROOFS_USE_MULTICORE_SDR=1 with 128GB RAM successfully and I don't know why.

With FIL_PROOFS_USE_MULTICORE_SDR=0, the same system is able to successfully perform two 32GiB PC1 tasks in parallel.

I've noticed the problem on Ryzen 3900X, 128GB RAM, 256GB swap, while sealing 32GiB sectors.

Also seen by @kernelogic (Slack Fei Yan - f010617) when sealing 64GB sectors, discussion: [https://github.com/filecoin-project/rust-fil-proofs/discussions/1409](https://github.com/filecoin-project/rust-fil-proofs/discussions/1409)

**Comment by cryptonemo** [comment]
Thanks for linking the discussion.  The mem mapping made is attempted to be locked.  If RAM and permissions align, it will be.  This is not compatible with swap, as the explicit purpose of mlock is to keep RAM in RAM and ensure it doesn't hit disk (via swap).

For probably most people, running as a non-root user will fail to lock the mapping and instead fallback to a non-mlocked mapping (works fine, and no noticeable performance is gained either way, at least according to observations by a few parties).

**Comment by kernelogic** [comment]
Aha, I indeed running my miner using root. So I guess I should change to non-root in order to not get OOM killed.

**Comment by neondragon** [comment]
In the linked discussion, I narrowed down the problem to unexpected behaviour in Linux 2.6.9 onwards. Running as the root user will ignore the ulimit -l 'max locked memory' setting, and instead allow the process to mlock an unlimited amount of memory.

Perhaps it would have been more intuitive as a separate setting e.g. FIL_PROOFS_LOCK_MEMORY=[0/1], default disabled, to avoid the situation where a machine works with FIL_PROOFS_USE_MULTICORE_SDR=0 but is OOM killed with FIL_PROOFS_USE_MULTICORE_SDR=1, and because using locked pages requires custom ulimit -l configuration anyway.

However, this could be addressed with improved documentation [https://github.com/filecoin-project/rust-fil-proofs/blob/master/README.md](../blob/master/README.md) which currently states

> Multicore SDR [...] is not enabled by default but can be activated by setting FIL_PROOFS_USE_MULTICORE_SDR=1.
> Best performance will also be achieved when it is possible to lock pages which have been memory-mapped. **This can be accomplished either by running the process as root**, or by increasing the system limit for max locked memory with ulimit -l.

That's why I was running lotus-bench as root.

I've submitted a pull request to perhaps clarify that a little [](https://github.com/filecoin-project/rust-fil-proofs/pull/1413)
